packages/edit-site/src/components/header/document-actions/index.js

```diff
@@ -25,45 +25,62 @@ function useSecondaryText() {
 		selectedBlock,
 		getBlockParentsByBlockName,
 		getBlockWithoutInnerBlocks,
+		hoveredTemplatePartBlock,
 	} = useSelect( ( select ) => {
+		const {
+			getSelectedBlock,
+			getBlockParentsByBlockName: _getBlockParentsByBlockName,
+			__unstableGetBlockWithoutInnerBlocks,
+			getHoveredBlockByBlockName,
+		} = select( 'core/block-editor' );
 		return {
-			selectedBlock: select( 'core/block-editor' ).getSelectedBlock(),
-			getBlockParentsByBlockName: select( 'core/block-editor' )
-				.getBlockParentsByBlockName,
-			getBlockWithoutInnerBlocks: select( 'core/block-editor' )
-				.__unstableGetBlockWithoutInnerBlocks,
+			selectedBlock: getSelectedBlock(),
+			getBlockParentsByBlockName: _getBlockParentsByBlockName,
+			getBlockWithoutInnerBlocks: __unstableGetBlockWithoutInnerBlocks,
+			hoveredTemplatePartBlock: getHoveredBlockByBlockName(
+				'core/template-part'
+			),
 		};
 	} );
 
 	// Check if current block is a template part:
-	const selectedBlockLabel =
+	let selectedBlockLabel =
 		selectedBlock?.name === 'core/template-part'
 			? getBlockDisplayText( selectedBlock )
 			: null;
 
-	if ( selectedBlockLabel ) {
+	// Check if an ancestor of the current block is a template part:
+	if ( ! selectedBlockLabel ) {
+		const templatePartParents = !! selectedBlock
+			? getBlockParentsByBlockName(
+					selectedBlock?.clientId,
+					'core/template-part'
+			  )
+			: [];
+
+		if ( templatePartParents.length ) {
+			// templatePartParents is in order from top to bottom, so the closest
+			// parent is at the end.
+			const closestParent = getBlockWithoutInnerBlocks(
+				last( templatePartParents )
+			);
+			selectedBlockLabel = getBlockDisplayText( closestParent );
+		}
+	}
+
+	if ( hoveredTemplatePartBlock ) {
+		const hoveredBlockLabel = getBlockDisplayText(
+			hoveredTemplatePartBlock
+		);
 		return {
-			label: selectedBlockLabel,
-			isActive: true,
+			label: hoveredBlockLabel,
+			isActive: hoveredBlockLabel === selectedBlockLabel,
```

Comment:
I think we should use client ID to compare. I could see a case where two labels might be the same (even though it'd be confusing)

Reply:
✅  Set it up to compare by clientId